# TO DO List

- [ ] Define functions for library imports, data import setup and other EDA steps already finalized, to help initialize the notebook every day without having to run every cell

- [ ] Confirm every Business Assumptions with the Community

# Library imports

** **

In [1]:
import pandas as pd
import numpy as np
import warnings
import seaborn as sns
import plotly.express as px
import scipy.stats as _stats
import datetime as dt
import itertools

from matplotlib import pyplot as plt
from IPython.core.display import display, HTML
from helper_functions import *

In [2]:
def jupyter_settings():
    %matplotlib inline
    %pylab inline
    
    plt.style.use( 'bmh' )
    plt.rcParams['figure.figsize'] = [20, 12]
    plt.rcParams['font.size'] = 24
    
    display( HTML( '<style>.container { width:100% !important; }</style>') )
    
    pd.options.display.max_columns = 200
    pd.options.display.max_rows = 200
    pd.set_option( 'display.expand_frame_repr', False )
    pd.set_option('display.float_format', lambda x: '%.2f' % x)
    
    sns.set()
    
    #warnings.filterwarnings("ignore")

In [3]:
jupyter_settings()

Populating the interactive namespace from numpy and matplotlib


/home/data-madness/Documents/SynologyDrive/Comunidade DS/Python do Zero ao DS/HouseSales-RecommendationSystem/houserocket_env/lib/python3.8/site-packages/IPython/core/magics/pylab.py:159: UserWarning: pylab import has clobbered these variables: ['save']
`%matplotlib` prevents importing * from pylab and numpy
  warn("pylab import has clobbered these variables: %s"  % clobbered +


** **

# Data Import and Setup

** **

### Import data
** **

In [4]:
orig_df = pd.read_csv('data/kc_house_data.csv')
house_df_eda = orig_df.copy(deep=True)

### General Info

** **

In [5]:
house_df_eda.head()

id             date     price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  view  condition  grade  sqft_above  sqft_basement  yr_built  yr_renovated  zipcode   lat    long  sqft_living15  sqft_lot15
0  7129300520  20141013T000000 221900.00         3       1.00         1180      5650    1.00           0     0          3      7        1180              0      1955             0    98178 47.51 -122.26           1340        5650
1  6414100192  20141209T000000 538000.00         3       2.25         2570      7242    2.00           0     0          3      7        2170            400      1951          1991    98125 47.72 -122.32           1690        7639
2  5631500400  20150225T000000 180000.00         2       1.00          770     10000    1.00           0     0          3      6         770              0      1933             0    98028 47.74 -122.23           2720        8062
3  2487200875  20141209T000000 604000.00         4       3.00         1960      5000    1.00           0     0          5      7        1050            910      1965             0    98136 47.52 -122.39           1360        5000
4  1954400510  20150218T000000 510000.00         3       2.00         1680      8080    1.00           0     0          3      8        1680              0      1987             0    98074 47.62 -122.05           1800        7503

In [6]:
house_df_eda.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

In [7]:
house_df_eda.describe().T

count          mean           std        min           25%           50%           75%           max
id            21613.00 4580301520.86 2876565571.31 1000102.00 2123049194.00 3904930410.00 7308900445.00 9900000190.00
price         21613.00     540088.14     367127.20   75000.00     321950.00     450000.00     645000.00    7700000.00
bedrooms      21613.00          3.37          0.93       0.00          3.00          3.00          4.00         33.00
bathrooms     21613.00          2.11          0.77       0.00          1.75          2.25          2.50          8.00
sqft_living   21613.00       2079.90        918.44     290.00       1427.00       1910.00       2550.00      13540.00
sqft_lot      21613.00      15106.97      41420.51     520.00       5040.00       7618.00      10688.00    1651359.00
floors        21613.00          1.49          0.54       1.00          1.00          1.50          2.00          3.50
waterfront    21613.00          0.01          0.09       0.00          0.00          0.00          0.00          1.00
view          21613.00          0.23          0.77       0.00          0.00          0.00          0.00          4.00
condition     21613.00          3.41          0.65       1.00          3.00          3.00          4.00          5.00
grade         21613.00          7.66          1.18       1.00          7.00          7.00          8.00         13.00
sqft_above    21613.00       1788.39        828.09     290.00       1190.00       1560.00       2210.00       9410.00
sqft_basement 21613.00        291.51        442.58       0.00          0.00          0.00        560.00       4820.00
yr_built      21613.00       1971.01         29.37    1900.00       1951.00       1975.00       1997.00       2015.00
yr_renovated  21613.00         84.40        401.68       0.00          0.00          0.00          0.00       2015.00
zipcode       21613.00      98077.94         53.51   98001.00      98033.00      98065.00      98118.00      98199.00
lat           21613.00         47.56          0.14      47.16         47.47         47.57         47.68         47.78
long          21613.00       -122.21          0.14    -122.52       -122.33       -122.23       -122.12       -121.31
sqft_living15 21613.00       1986.55        685.39     399.00       1490.00       1840.00       2360.00       6210.00
sqft_lot15    21613.00      12768.46      27304.18     651.00       5100.00       7620.00      10083.00     871200.00

** **

# 1. Exploratory Data Analysis
** **

### 1.1 Data Dimensionality
** **

It is a small sized dataset, and and it will not suffer from the curse of dimensionality.

In [8]:
print('Number of Rows:{}'.format(house_df_eda.shape[0]))
print('Number of Columns {}'.format(house_df_eda.shape[1]))

Number of Rows:21613
Number of Columns 21


### 1.2 Attributes Name and Description
** **

**id** - its a unique identifier number for each unique house

**date** - its the date when the house were sold

**price** - its the selling price when the house was sold

**bedrooms** - number of bedrooms in the house

**bathrooms** - number of bathrooms in the house, where a fraction like 0.25 represents a bathroom sink, shower or toilet

**sqft_living** - square footage of the apartments interior living space

**sqft_lot** - the size of the land in square feet

**floors** - number of floors in the house

**waterfront** - if there is a waterview from the house

**view** - an index from 0 to 4 of how good the view of the property was or how many views the house has.

**condition** - the house preservation condition

**grade** - a rank from 1 to 13, which ranks the construction quality

**sqft_above** - the size of the house above the ground level in square feet

**sqft_basement** - the size of the house below the ground level in square feet

**yr_built** - the year the house was initially built

**yr_renovated** - the year of the house's last renovation

**zipcode** - what zipcode area the house is in

**lat** - Lattitude

**long** - Longitude

**sqft_living15** - The square footage of interior housing living space for the nearest 15 neighbors (possibly)

**sqft_lot15** - The square footage of the land lots of the nearest 15 neighbors

** **

### 1.3 Attributes Type and Data type conversions
** **

In [9]:
house_df_eda.dtypes

id                 int64
date              object
price            float64
bedrooms           int64
bathrooms        float64
sqft_living        int64
sqft_lot           int64
floors           float64
waterfront         int64
view               int64
condition          int64
grade              int64
sqft_above         int64
sqft_basement      int64
yr_built           int64
yr_renovated       int64
zipcode            int64
lat              float64
long             float64
sqft_living15      int64
sqft_lot15         int64
dtype: object

>#### Date reformat type

Because it will be handled easily as a date variable type

In [10]:
house_df_eda.date = pd.to_datetime(house_df_eda.date)

In [11]:
house_df_eda.date[0]

Timestamp('2014-10-13 00:00:00')

### 1.4 Attributes Descriptive Statistics (Summary Statistics)
** **

In [12]:
num_attributes = house_df_eda.select_dtypes(include=['int64', 'float64'])

In [13]:
num_attributes.columns

Index(['id', 'price', 'bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot',
       'floors', 'waterfront', 'view', 'condition', 'grade', 'sqft_above',
       'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode', 'lat', 'long',
       'sqft_living15', 'sqft_lot15'],
      dtype='object')

In [14]:
mean = pd.DataFrame(num_attributes.mean())
std = pd.DataFrame(num_attributes.std())
median = pd.DataFrame(num_attributes.median())

q1,q3 = pd.DataFrame(num_attributes.quantile([0.25])), pd.DataFrame(num_attributes.quantile([0.75]))

maximum = pd.DataFrame(num_attributes.max())
minimum = pd.DataFrame(num_attributes.min())

skew = pd.DataFrame(num_attributes.skew())
kurtosis = pd.DataFrame(num_attributes.kurtosis())

descriptive_statistics = pd.concat([mean,std,minimum,q1.transpose(),median,q3.transpose(),maximum,skew,kurtosis],axis=1)
descriptive_statistics.columns = ['mean','std','minimum','q1','median','q3','maximum','skew','kurtosis']
descriptive_statistics

mean           std    minimum            q1        median            q3       maximum  skew  kurtosis
id            4580301520.86 2876565571.31 1000102.00 2123049194.00 3904930410.00 7308900445.00 9900000190.00  0.24     -1.26
price             540088.14     367127.20   75000.00     321950.00     450000.00     645000.00    7700000.00  4.02     34.59
bedrooms               3.37          0.93       0.00          3.00          3.00          4.00         33.00  1.97     49.06
bathrooms              2.11          0.77       0.00          1.75          2.25          2.50          8.00  0.51      1.28
sqft_living         2079.90        918.44     290.00       1427.00       1910.00       2550.00      13540.00  1.47      5.24
sqft_lot           15106.97      41420.51     520.00       5040.00       7618.00      10688.00    1651359.00 13.06    285.08
floors                 1.49          0.54       1.00          1.00          1.50          2.00          3.50  0.62     -0.48
waterfront             0.01          0.09       0.00          0.00          0.00          0.00          1.00 11.39    127.63
view                   0.23          0.77       0.00          0.00          0.00          0.00          4.00  3.40     10.89
condition              3.41          0.65       1.00          3.00          3.00          4.00          5.00  1.03      0.53
grade                  7.66          1.18       1.00          7.00          7.00          8.00         13.00  0.77      1.19
sqft_above          1788.39        828.09     290.00       1190.00       1560.00       2210.00       9410.00  1.45      3.40
sqft_basement        291.51        442.58       0.00          0.00          0.00        560.00       4820.00  1.58      2.72
yr_built            1971.01         29.37    1900.00       1951.00       1975.00       1997.00       2015.00 -0.47     -0.66
yr_renovated          84.40        401.68       0.00          0.00          0.00          0.00       2015.00  4.55     18.70
zipcode            98077.94         53.51   98001.00      98033.00      98065.00      98118.00      98199.00  0.41     -0.85
lat                   47.56          0.14      47.16         47.47         47.57         47.68         47.78 -0.49     -0.68
long                -122.21          0.14    -122.52       -122.33       -122.23       -122.12       -121.31  0.89      1.05
sqft_living15       1986.55        685.39     399.00       1490.00       1840.00       2360.00       6210.00  1.11      1.60
sqft_lot15         12768.46      27304.18     651.00       5100.00       7620.00      10083.00     871200.00  9.51    150.76

### 1.5 Identify missing values and Duplicated records
***

>#### Missing Values

In [15]:
num_attributes.isnull().sum()

id               0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64

In [16]:
cat_attributes = house_df_eda.select_dtypes(exclude=['int64','float64'])

pd.isna(cat_attributes).sum()

date    0
dtype: int64

>#### Check Duplicates (Houses sold twice or more times)

In [17]:
len(house_df_eda['id'].unique())

21436

In [18]:
duplicated_ids = house_df_eda['id'].apply(lambda x: True if (sum(x == house_df_eda['id']) >= 2)\
                                                     else False)

Drop the duplicates but keep the most recent, because for this business purpose it is relevant the houses that are currently on market or the real estate portfolio.

In [19]:
house_df_eda.loc[duplicated_ids,['id','date','price', 'yr_renovated','zipcode']].drop_duplicates(subset=['id'], keep='last')

id       date      price  yr_renovated  zipcode
94     6021501535 2014-12-23  700000.00             0    98117
314    4139480200 2014-12-09 1400000.00             0    98006
325    7520000520 2015-03-11  240500.00          1984    98146
346    3969300030 2014-12-29  239900.00             0    98178
372    2231500030 2015-03-24  530000.00             0    98133
718    8820903380 2015-01-02  730000.00          1990    98125
824     726049190 2015-02-18  431000.00             0    98133
837    8682262400 2015-05-13  419950.00             0    98053
1086   9834200885 2015-04-20  550000.00             0    98144
1129   8062900070 2015-02-13  369000.00             0    98056
1203   1231000510 2015-05-04  510000.00             0    98118
1235   6117501820 2015-04-28  435000.00             0    98166
1451   2228900270 2015-02-12  302000.00             0    98133
1465   3630120700 2015-01-07  765000.00             0    98029
1577   7888000390 2015-04-01  235000.00             0    98198
1865   2143700830 2015-03-12  370000.00             0    98055
2039   3395040550 2015-04-29  320000.00             0    98108
2127   4345000510 2015-04-28  325000.00             0    98030
2494   9407110710 2015-02-26  322000.00             0    98045
2497      1000102 2015-04-22  300000.00             0    98002
2503   1781500435 2015-02-25  575000.00             0    98126
2532   1232000810 2015-03-26  537000.00             0    98117
2565   7856400240 2015-02-11 1650000.00             0    98006
2632   7983000200 2015-02-25  250000.00             0    98003
2977   9136103130 2015-05-12  685000.00             0    98103
3034   5083000375 2015-03-19  235000.00             0    98198
3299   7409700215 2015-03-13  921500.00             0    98115
3541   1823049202 2015-01-07  326000.00             0    98146
3624   2767602141 2014-12-22  650000.00             0    98107
3757   1254200015 2015-04-08  625000.00             0    98117
3786    723049156 2014-11-12  284700.00             0    98146
3794   4435000705 2015-03-09  255500.00             0    98188
3879   4364700600 2015-03-30  390000.00             0    98126
3951   1825069031 2014-10-16  550000.00          1980    98074
4078   3262300940 2015-02-10  940000.00             0    98039
4343   9353300600 2015-03-26  370000.00             0    98059
4873   1524079093 2015-03-18  369500.00             0    98024
4923   9809000020 2015-03-13 1940000.00             0    98004
5245   3303000130 2015-04-28  520000.00             0    98177
5252   7524400250 2014-11-24  589950.00             0    98052
5273   4204400098 2015-04-21  385000.00             0    98055
5341   6632900574 2015-02-25  595000.00             0    98155
5593   2726049071 2015-04-08  489950.00             0    98125
5695   3578401060 2015-05-04  625000.00             0    98074
5724   1788800630 2015-02-25  185000.00             0    98023
5757   6381500170 2015-01-16  365000.00             0    98125
5759   1954420170 2014-11-13  580000.00             0    98074
5973   1545800290 2015-04-08  315000.00             0    98038
6346   9828200460 2015-01-06  430000.00             0    98122
6372   9222400605 2015-04-11  850000.00             0    98115
6435   8832900780 2015-04-08  647500.00             0    98028
6534   8645530010 2015-03-25  295000.00             0    98058
6631   3323059027 2015-02-25  340000.00             0    98058
6720   5417600130 2015-05-12  301000.00             0    98065
6790   8651510380 2014-12-16  539000.00             0    98074
6903   5249801440 2015-04-22  570000.00             0    98118
7179    526059224 2015-02-06  470000.00             0    98011
7246   1446403850 2015-01-14  212000.00             0    98168
7793    251300110 2015-01-14  358000.00             0    98003
7846   4139420590 2014-08-27 1200000.00             0    98006
7848   2621600015 2015-04-30  175000.00             0    98030
8012   2206700215 2015-04-22  550000.00             0    98006
8126   2787460720 2015-05-06  259950.00       

### 1.6 Data Granularity
***

>#### Geographic Data

The geographic data attributes *latitude* and *longitude* are in their finest granularity, but *zipcode* despite being in their purest from in terms of value, it could be transformed into other finer variables to describe specifically the location of the house, for example, the street or a more coarse one as county. 

>#### Temporal Data

The data attributes *yr_built* and *yr_renovated* are interval-based and nominal, respectively. And both have an year level of granularity. The *date* attribute can be considered as interval-based with a day level of temporal granularity

In [20]:
# house_df_eda['yr_built'].sort_values().unique()

In [21]:
# house_df_eda['yr_renovated'].sort_values().unique()

In [22]:
# house_df_eda['date'].sort_values().unique()

>#### Numerical Data 

All numeric variables {'id', 'price', 'bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot',
       'floors', 'waterfront', 'view', 'condition', 'grade', 'sqft_above',
       'sqft_basement', 'sqft_living15', 'sqft_lot15'}
      
have the finest granularity, they are not composed or discretized in any way.

### 1.7 Data Distribution
***

In [23]:
# values = num_attributes['sqft_living'].sort_values().values
# tuplex = plt.hist(values, 20, density=True)

In [24]:
# rows, cols = choose_grid(len(num_attributes.columns))

# iterate_through_data(num_attributes,num_attributes.columns,rows,cols)

In [25]:
# identify_best_fit_distribution(num_attributes,'price')

In [26]:
# identify_best_fit_distribution(num_attributes,'sqft_lot')

### 1.8 Data Sparsity
***

In [27]:
# multiple_scatter_plots(house_df_eda.loc[:, house_df_eda.columns != 'id'])

> #### Correlation Analysis

It would be important to do some correlation analysis with the derived measures in the second cycle.

The *sqft_living* and *sqft_above* have an obvious high positive correlation.

Both *sqft_living* and *sqft_above* have an high positive correlation with the *grade*. So for the renovation question, it would be interesting to see if increasing *sqft_living* the selling price would also increase.

In [28]:
# fig = plt.figure(figsize=[12, 12])
# corr_mtx = num_attributes.corr()
# sns.heatmap(corr_mtx, xticklabels=corr_mtx.columns, yticklabels=corr_mtx.columns, annot=True, cmap='Blues')
# plt.title('Correlation analysis')
# plt.show()

> #### Outlier Analysis

Outliers still not analyzed

In [29]:
# Outlier identification functions

def z_score(data, k_dev):
    mean = np.round(data.mean(), decimals=2)
    std_dev = np.round(data.std(), decimals=2)

    z_scores = [ (x - mean)/std_dev for x in data ]
    return data[(np.abs(z_scores) > k_dev)]

def modified_zscore(data, thresh):

    median = np.median(data)
    median_absolute_deviation = np.median([np.abs(x - median) for x in data])

    modified_z_scores = [0.6745 * (x - median) / median_absolute_deviation for x in data]
    return data[(np.abs(modified_z_scores) > thresh)]

def iqr(data, dist):
  
    Q1 = data.quantile(0.25)
    Q3 = data.quantile(0.75)
    IQR = Q3 - Q1

    return data[(data < (Q1 - (dist * IQR))) | (data > (Q3 + (dist * IQR)))]

In [30]:
house_df_eda.loc[:5, ~house_df_eda.columns.isin(['id','date'])]

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  view  condition  grade  sqft_above  sqft_basement  yr_built  yr_renovated  zipcode   lat    long  sqft_living15  sqft_lot15
0  221900.00         3       1.00         1180      5650    1.00           0     0          3      7        1180              0      1955             0    98178 47.51 -122.26           1340        5650
1  538000.00         3       2.25         2570      7242    2.00           0     0          3      7        2170            400      1951          1991    98125 47.72 -122.32           1690        7639
2  180000.00         2       1.00          770     10000    1.00           0     0          3      6         770              0      1933             0    98028 47.74 -122.23           2720        8062
3  604000.00         4       3.00         1960      5000    1.00           0     0          5      7        1050            910      1965             0    98136 47.52 -122.39           1360        5000
4  510000.00         3       2.00         1680      8080    1.00           0     0          3      8        1680              0      1987             0    98074 47.62 -122.05           1800        7503
5 1225000.00         4       4.50         5420    101930    1.00           0     0          3     11        3890           1530      2001             0    98053 47.66 -122.00           4760      101930

In [31]:
outliers = {}
for feature in num_attributes.loc[:, ~num_attributes.columns.isin(['id','date'])].columns:
#     print(f"{feature} outliers", end="\n")
#     print(iqr(num_attributes[feature], 1.5))
    outliers[feature] = iqr(num_attributes[feature], 1.5)

In [32]:
outliers['bedrooms'].unique()

array([ 1,  6,  7,  0,  8,  9, 11, 10, 33])

### 1.9 Identify Data Inconsistency
***

Some inconsitent records or records with inconsistent values must be removed if they are not relevant outliers, because they will bias the analysis and future model training.


In [33]:
house_df_eda_inconsistency = house_df_eda.copy(deep=True)

In [34]:
house_df_eda_inconsistency.loc[house_df_eda_inconsistency['bedrooms'] == 0]

id       date      price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  view  condition  grade  sqft_above  sqft_basement  yr_built  yr_renovated  zipcode   lat    long  sqft_living15  sqft_lot15
875    6306400140 2014-06-12 1095000.00         0       0.00         3064      4764    3.50           0     2          3      7        3064              0      1990             0    98102 47.64 -122.32           2360        4000
3119   3918400017 2015-02-05  380000.00         0       0.00         1470       979    3.00           0     2          3      8        1470              0      2006             0    98133 47.71 -122.36           1470        1399
3467   1453602309 2014-08-05  288000.00         0       1.50         1430      1650    3.00           0     0          3      7        1430              0      1999             0    98125 47.72 -122.29           1430        1650
4868   6896300380 2014-10-02  228000.00         0       1.00          390      5900    1.00           0     0          2      4         390              0      1953             0    98118 47.53 -122.26           2170        6000
6994   2954400190 2014-06-24 1295650.00         0       0.00         4810     28008    2.00           0     0          3     12        4810              0      1990             0    98053 47.66 -122.07           4740       35061
8477   2569500210 2014-11-17  339950.00         0       2.50         2290      8319    2.00           0     0          3      8        2290              0      1985             0    98042 47.35 -122.15           2500        8751
8484   2310060040 2014-09-25  240000.00         0       2.50         1810      5669    2.00           0     0          3      7        1810              0      2003             0    98038 47.35 -122.05           1810        5685
9773   3374500520 2015-04-29  355000.00         0       0.00         2460      8049    2.00           0     0          3      8        2460              0      1990             0    98031 47.41 -122.17           2520        8050
9854   7849202190 2014-12-23  235000.00         0       0.00         1470      4800    2.00           0     0          3      7        1470              0      1996             0    98065 47.53 -121.83           1060        7200
12653  7849202299 2015-02-18  320000.00         0       2.50         1490      7111    2.00           0     0          3      7        1490              0      1999             0    98065 47.53 -121.83           1500        4675
14423  9543000205 2015-04-13  139950.00         0       0.00          844      4269    1.00           0     0          4      7         844              0      1913             0    98001 47.28 -122.25           1380        9600
18379  1222029077 2014-10-29  265000.00         0       0.75          384    213444    1.00           0     0          3      4         384              0      2003             0    98070 47.42 -122.49           1920      224341
19452  3980300371 2014-09-26  142000.00         0       0.00          290     20875    1.00           0     0          1      1         290              0      1963             0    98024 47.53 -121.89           1620       22850

In [35]:
house_df_eda_inconsistency.loc[house_df_eda_inconsistency['bedrooms'] == 33]

id       date     price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  view  condition  grade  sqft_above  sqft_basement  yr_built  yr_renovated  zipcode   lat    long  sqft_living15  sqft_lot15
15870  2402100895 2014-06-25 640000.00        33       1.75         1620      6000    1.00           0     0          5      7        1040            580      1947             0    98103 47.69 -122.33           1330        4700

In [36]:
house_df_eda_inconsistency.yr_renovated.sort_values().unique()

array([   0, 1934, 1940, 1944, 1945, 1946, 1948, 1950, 1951, 1953, 1954,
       1955, 1956, 1957, 1958, 1959, 1960, 1962, 1963, 1964, 1965, 1967,
       1968, 1969, 1970, 1971, 1972, 1973, 1974, 1975, 1976, 1977, 1978,
       1979, 1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987, 1988, 1989,
       1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000,
       2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011,
       2012, 2013, 2014, 2015])

In [37]:
len(house_df_eda_inconsistency.loc[house_df_eda_inconsistency['yr_renovated'] == 0])

20699

In [38]:
house_df_eda_inconsistency['living_area_above_and_basement'] = house_df_eda_inconsistency['sqft_living'] - (house_df_eda_inconsistency['sqft_above'] + house_df_eda_inconsistency['sqft_basement'])

In [39]:
house_df_eda_inconsistency.loc[house_df_eda_inconsistency['living_area_above_and_basement'] < 0]

Empty DataFrame
Columns: [id, date, price, bedrooms, bathrooms, sqft_living, sqft_lot, floors, waterfront, view, condition, grade, sqft_above, sqft_basement, yr_built, yr_renovated, zipcode, lat, long, sqft_living15, sqft_lot15, living_area_above_and_basement]
Index: []

#### Inconsistent features values

1. bedrooms
    1. There are some houses with 0 bedrooms, but may be houses with no purpose of living.
    
    2. There is some house record with 33 bedrooms with only 1040 square feet of infraestructure size


2. bathrooms
    1. There are some houses with bathrooms variable with decimal places, like 1.75, that could be inconsistency, but it is assumed each 1 unit of bathroom is a complete bathroom with sink, shower or toilet, each one being 0.25.


3. yr_renovated
    1. 20699 records have the year renovated variable equal to 0
    
    
4. sqft_living - (sqft_above - sqft_basement)

    It wasn't found any inconsistency

#### Inconsistent records

### 1.10 Business Data Assumptions
***

It is important to do some business data assumption to better analyze the data and take interpretable insights.

1. **Region granularity**

    There is some difference between sub-regions in King County, so there must be a separation between regions, and have a specific analysis for each. Since a generalized analysis would nullify the correlation of each region, and would be harder to take insights. This region segmentation can be made in different granularities, that is it can be divided by North, South, and East Mountains, but it can also have a finer granularity that is for zipcode regions. Another finer granularity can be the street or street block by collecting that data from the latitude and longitude, and counting the number of houses per region to assess if there is enough data for each street to get significant statistical data.


2. **Bedrooms inconsistency**

    Houses with 0 bedrooms, may be for other purposes apart from living usage. (Confirm all records consistency, analysing one by one or doing some summary statistics)
    
    
3. **Year Renovated inconsistency**

    It can be assumed that houses with value 0 on yr_renovated is basically houses that have not been renovated yet.
    
    
4. **Condition vs Grade relevance**
    
    The condition variable it seems to have a higher weight in long-term than grade, in terms of evaluating
    which houses are better to buy, since the grade is the infraestructure construction quality and it cannot
    change so much with time, unlike condition that is the house preservation and it can it can increase with
    some maintenance works.

### 1.12 Attributes Usefulness for the tasks
***

First intuition without looking to EDA, and only based on the business understanding. (On the second cycle will be based on the EDA) 

**Business:** Buy and sell houses in Real Estate market, with some profit.

**Business problems:**

    1. Which houses should be bought and for what price?
    
    2. Once its bought when it's the best time period to sell it and for what price?

    3. To rise the housing selling price, the company should do a renovation. So what would be good renewal changes?

    This tasks can easily be done as a Classification Task, after doing the labelling task of giving a label to each record, telling if the house should be bought or not. This labelling process comes with some business assumptions of good profit values and empirical rules, and by doing a preliminary analysis we get our first actionable insights that will be presented to business experts. After validating this labelling empirical process, we can train a classification model, to make the decision-making more robust and less biased by the history.
    
    Useful Attributes for business question:
    
    1. date, price, condition, grade, zipcode
    2. date, price, condition, grade, zipcode
    3. date, price, bedrooms, bathrooms, sqft_living, sqft_above

### 1.13 Feature Selection
***

In [69]:
house_df_fs = house_df_eda[['id','date','price','condition','grade','zipcode','bedrooms','bathrooms','sqft_living','sqft_above','sqft_basement','sqft_lot']]

### 1.14 Feature engineering and derivation (Identify extra useful features based on solution planning)
***

(Check planning document)

Useful Attributes for business question:
    
    1. date, price, condition, grade, zipcode, plus:
        1. median_price - based on zipcode region of the selected houses
        2. percentage_value_below_median
        3. house_total_m2 - m2_living + m2_lot
        4. price/house_total_m2 - will help more obtaining more informative insights on comparisons, since we are normalizing the 
        price by the house size, and then there is a more fair comparison.
    2. From the selected houses to buy create and use:
        1. best_season - based on zipcode region of the selected houses, and its selling date
        2. selling_price - based on the price and the season
        3. profit - will result from difference between selling_price and price 
    3. From the selected houses to buy create and use:
        1.
        2. 

In [70]:
house_df_fs[['m2_living','m2_above','m2_basement','m2_lot']] = house_df_fs[['sqft_living','sqft_above','sqft_basement','sqft_lot']] * 0.0929

house_df_fs.drop(['sqft_living','sqft_above','sqft_basement','sqft_lot'], axis=1, inplace=True)

house_df_fs.head()

/home/data-madness/Documents/SynologyDrive/Comunidade DS/Python do Zero ao DS/HouseSales-RecommendationSystem/houserocket_env/lib/python3.8/site-packages/pandas/core/frame.py:3641: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[k1] = value[k2]
/home/data-madness/Documents/SynologyDrive/Comunidade DS/Python do Zero ao DS/HouseSales-RecommendationSystem/houserocket_env/lib/python3.8/site-packages/pandas/core/frame.py:4906: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


id       date     price  condition  grade  zipcode  bedrooms  bathrooms  m2_living  m2_above  m2_basement  m2_lot
0  7129300520 2014-10-13 221900.00          3      7    98178         3       1.00     109.62    109.62         0.00  524.88
1  6414100192 2014-12-09 538000.00          3      7    98125         3       2.25     238.75    201.59        37.16  672.78
2  5631500400 2015-02-25 180000.00          3      6    98028         2       1.00      71.53     71.53         0.00  929.00
3  2487200875 2014-12-09 604000.00          5      7    98136         4       3.00     182.08     97.55        84.54  464.50
4  1954400510 2015-02-18 510000.00          3      8    98074         3       2.00     156.07    156.07         0.00  750.63

In [71]:
house_df_fs['house_total_m2'] = house_df_fs['m2_living'] + house_df_fs['m2_lot']

house_df_fs.head()

/tmp/ipykernel_16988/3069127675.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  house_df_fs['house_total_m2'] = house_df_fs['m2_living'] + house_df_fs['m2_lot']


id       date     price  condition  grade  zipcode  bedrooms  bathrooms  m2_living  m2_above  m2_basement  m2_lot  house_total_m2
0  7129300520 2014-10-13 221900.00          3      7    98178         3       1.00     109.62    109.62         0.00  524.88          634.51
1  6414100192 2014-12-09 538000.00          3      7    98125         3       2.25     238.75    201.59        37.16  672.78          911.53
2  5631500400 2015-02-25 180000.00          3      6    98028         2       1.00      71.53     71.53         0.00  929.00         1000.53
3  2487200875 2014-12-09 604000.00          5      7    98136         4       3.00     182.08     97.55        84.54  464.50          646.58
4  1954400510 2015-02-18 510000.00          3      8    98074         3       2.00     156.07    156.07         0.00  750.63          906.70

## 2. Find Main Insights

### 2.1 House Recommendation report (Answering Business Team questions)

### 2.1.1 Which houses should be bought and for what price?

#### Business Assumptions
***

The recommendation solution will output houses to buy and not to buy only based on the region median price, because residence localization is an important business information, and more specifically based on the median price of that region but taking into account the property size, since its an important house comparison metric, because the price of a house with  60m² that is below the median price of its region does not tell that it is a potential house to buy and sell for a higher price near this region median, since the region property size distribution may be left skewed and the majority contribution for that median price comes from houses with bigger property size. 

So, it is better to use a metric like price/m2 to do a fair evaluation of the real estate or the property size appreciation for each region, that can be influenced by the mobility access of that region, and what offers in terms of market and public spaces. Denoting that this decision making process will be based on the infraestructure construction area that is the aspect of the house that would remotely be changed, unlike some maintenance or renovation works to preserve the house which could be what could be improved to increase that evaluation. 

The price value will only be used to calculate the profit, coming from the difference with the selling price obtained through an increase on appreciation price/m2 by using as baseline the current median price/m2, that always suffer some market inflation, because the works expenses always increase each year and consequently other services and the house value, ignoring the possible increase with market appreciation because of higher house demand and less offer.

The decision to buy the houses is fundamentally based on the potential profit based on the expenses and the potential margin to increase value to match the median price/m2 of the respective region. The houses not considered to buy, can have different evaluation afterwards.

Why using median price and not mean?

The mean is usually influenced by higher values, so depending on the identities attribute distribution, e.g. the houses pricing value distribution, it may be a bad metric to make decisions upon this baseline. Instead we will use median, that is less prone to follow bigger values. 

**1. Identify best houses with comparison within each region**

**-** The houses that have a price value lower than the median price/m2, compared with houses that belong to the same region, and are in good conditions (condition $\epsilon$ [3,5]), can be sold for a higher price, so are good to buy.
    
    
**-** The houses that have a price value lower than the median price/m2, compared with houses that belong to the same region, and are in bad conditions (condition $\epsilon$ [0,2]), cannot be sold for a higher price, so are not so good to buy.

**-** The houses that have a price value higher than the median price/m2, compared with houses that belong to the same region, independently from the condition, are not good to buy to make good profit, but are going to be good for future comparison.

**-** The condition rank 3 takes into consideration that houses from a region that are below the region median price/m2 but has an average condition it's better than low condition houses that would need more renovation and could not have profit by taking into account the expenses, and has more potential to be sold above the median price of the region and get a higher profit since it has more margin for renewal and maintenance to increase the preservation condition unlike the houses with already good conditions.  


**2. Define a selling price estimation**

**-** Assuming that the houses to buy have preservation condition rank between 3 and 5, there is some expectancy to have more profit gradually from houses with rank 3 to 5, on descending order, since the houses with rank 5 may have been sold below the region median because of bad real estate marketing strategy, or it was sold on a time period of low appreciation, that will be studied on the next business question, and the rank 3 were more probable sold below that appreciation value because of obvious lower preservation conditions and few amenities. 

**-** So we set some different levels of increasing expectancy on houses appreciation:

    1. giving a 15% increase estimation on selling price, with region median price/m2 as baseline to calculate the current market price for houses with this property size and then adding the inflation, for houses with condition rank 3,
    2. 10% for houses with condition rank 4
    3. and 5% for houses with condition rank 5
    
**3. Ordering houses to buy with defined priorities**

    1. Order by percentage_below_median_price/m2, on a descending order
    2. then by condition, on a ascending order
    3. then by profit, on a descending order
    
    I'm giving priority to the percentage value below the median price/m² because it is the metric that gives more confidence since it has less derivation dependency, i.e. it results directly from measuring the proportion of the price/m², unlike profit that is based on a selling price derivation, and the selling price is found based on empirical business assumptions.

#### 2.1.1.1 Houses to buy labelling process

Labels:

    - 'to buy'
    - 'to compare'
    - 'not worth buying'

In [72]:
house_df_fs['condition'].value_counts()

3    14031
4     5679
5     1701
2      172
1       30
Name: condition, dtype: int64

In [73]:
# Create price/m2 and calculate its median per region.

house_df_fs['price_m2'] = house_df_fs['price']/house_df_fs['house_total_m2']

zipcode_median_price_m2 = house_df_fs[['price_m2', 'zipcode']].groupby('zipcode').median().reset_index()
zipcode_median_price_m2.columns = ['zipcode','median_price_m2']

house_df_fs = pd.merge(house_df_fs,zipcode_median_price_m2,on='zipcode',how='inner')

/tmp/ipykernel_16988/1218641239.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  house_df_fs['price_m2'] = house_df_fs['price']/house_df_fs['house_total_m2']


In [88]:
house_df_fs[house_df_fs['id'] == 2623069031]

id       date     price  condition  grade  zipcode  bedrooms  bathrooms  m2_living  m2_above  m2_basement   m2_lot  house_total_m2  price_m2  median_price_m2  status  selling_price  %_value_below_median_price_m2  profit_est
8361  2623069031 2014-05-21 542500.00          5      8    98027         5       3.25     279.63    186.73        92.90 99794.85       100074.48      5.42           485.63  to buy    51029226.15                          98.88 50486726.15

In [74]:
# Create house recommendation system

house_df_fs['status'] = house_df_fs[['price_m2','condition','median_price_m2']].apply(lambda x: 'to buy' if (x[0] < x[2]) & (x[1] >= 3)\
                                                                                           else 'to compare' if (x[0] > x[2])\
                                                                                           else 'not worth buying',  axis = 1)

In [75]:
house_df_fs[['id','zipcode','price','price_m2','median_price_m2','condition','status']].head()

id  zipcode     price  price_m2  median_price_m2  condition      status
0  7129300520    98178 221900.00    349.72           344.10          3  to compare
1  4060000240    98178 205425.00    288.67           344.10          4      to buy
2  4058801670    98178 445000.00    465.01           344.10          3  to compare
3  2976800796    98178 236000.00    352.93           344.10          3  to compare
4  6874200960    98178 170000.00    298.76           344.10          3      to buy

In [76]:
house_df_fs.status.value_counts()

to compare          10791
to buy              10623
not worth buying      199
Name: status, dtype: int64

#### 2.1.1.2 Create a selling price business model


In [77]:
# set a selling price

house_df_fs['selling_price'] = house_df_fs[['median_price_m2','house_total_m2','condition','status','price']].apply(lambda x: (x[0]*x[1])*1.15 if (x[2] == 3) & (x[3] == 'to buy')\
                                                                                    else (x[0]*x[1])*1.10 if (x[2] == 4) & (x[3] == 'to buy')\
                                                                                    else (x[0]*x[1])*1.05 if (x[2] == 5) & (x[3] == 'to buy')\
                                                                                    else x[4], axis=1)

In [78]:
house_df_fs[['id','zipcode','price','price_m2','median_price_m2','condition','status','selling_price']].head()

id  zipcode     price  price_m2  median_price_m2  condition      status  selling_price
0  7129300520    98178 221900.00    349.72           344.10          3  to compare      221900.00
1  4060000240    98178 205425.00    288.67           344.10          4      to buy      269350.01
2  4058801670    98178 445000.00    465.01           344.10          3  to compare      445000.00
3  2976800796    98178 236000.00    352.93           344.10          3  to compare      236000.00
4  6874200960    98178 170000.00    298.76           344.10          3      to buy      225164.27

#### 2.1.1.3 Add a priority order for selected houses to buy

In [79]:
# create percentage_value_below_median

house_df_fs['%_value_below_median_price_m2'] = (1 - (house_df_fs['price_m2']/house_df_fs['median_price_m2']))*100

In [80]:
# Create profit variable

house_df_fs['profit_est'] = house_df_fs['selling_price'] - house_df_fs['price']

In [81]:
house_df_fs.sort_values(by=['%_value_below_median_price_m2', 'condition', 'profit_est'], ascending=[False,True,False], inplace=True)

**!!! ->** Verify the price_m2 and median_price_m2 actual values

In [82]:
house_df_fs.loc[(house_df_fs['%_value_below_median_price_m2'] > 0) & (house_df_fs['status'] == 'to buy'),['id','zipcode','price','price_m2','median_price_m2','%_value_below_median_price_m2','condition','status','selling_price','profit_est']].head(20)

id  zipcode     price  price_m2  median_price_m2  %_value_below_median_price_m2  condition  status  selling_price  profit_est
8361   2623069031    98027 542500.00      5.42           485.63                          98.88          5  to buy    51029226.15 50486726.15
17671   123079023    98065 356000.00     10.43           650.87                          98.40          3  to buy    25542867.16 25186867.16
3383   1222069089    98038 375000.00      7.55           399.78                          98.11          5  to buy    20839957.03 20464957.03
18781  1020069017    98022 700000.00      4.56           227.53                          98.00          4  to buy    38426306.82 37726306.82
4858   1549500370    98019 210000.00      7.33           336.81                          97.82          3  to buy    11089479.59 10879479.59
9295    522069097    98058 150000.00      7.59           314.84                          97.59          3  to buy     7159735.45  7009735.45
3185   3323069045    98038 234000.00     10.48           399.78                          97.38          3  to buy    10266829.31 10032829.31
17578  3124089086    98065 300000.00     18.00           650.87                          97.23          3  to buy    12473820.32 12173820.32
18794  3520069033    98022 230000.00      6.34           227.53                          97.21          4  to buy     9083221.23  8853221.23
1996   1525069021    98053 400000.00     19.85           671.28                          97.04          3  to buy    15554813.01 15154813.01
3352    522079022    98038 700000.00     12.03           399.78                          96.99          4  to buy    25586700.26 24886700.26
3437   1422069070    98038 472000.00     12.18           399.78                          96.95          3  to buy    17809521.80 17337521.80
1940   1525069088    98053 442500.00     20.68           671.28                          96.92          3  to buy    16518817.69 16076317.69
2004   2325069054    98053 225000.00     21.37           671.28                          96.82          3  to buy     8128634.96  7903634.96
6831    120059044    98092 250000.00      9.34           289.71                          96.77          3  to buy     8913410.61  8663410.61
18878  3420069055    98022 350000.00      7.52           227.53                          96.70          4  to buy    11652130.19 11302130.19
8551   2623069010    98027 745000.00     16.09           485.63                          96.69          5  to buy    23602724.81 22857724.81
18795  1820069019    98022 302000.00      7.65           227.53                          96.64          5  to buy     9426774.64  9124774.64
1824   1525069058    98053 568000.00     22.83           671.28                          96.60          4  to buy    18372270.63 17804270.63
17669  2024089011    98065 550000.00     22.39           650.87                          96.56          3  to buy    18383946.93 17833946.93

In [85]:
recommended_houses_ordered = house_df_fs.loc[(house_df_fs['%_value_below_median_price_m2'] > 0) & (house_df_fs['status'] == 'to buy'),:].copy(deep=True)

In [86]:
recommended_houses_ordered.shape[0]

10623

**Next Step**

A seguir a obter as casas finais deveria fazer um mapa com a distribuição das mesmas com a cor pela condition e o tamanho pelo valor de mercado (última venda) para assinalar as mais caras e baratas de adquirir. Depois um mapa de densidade pela variável profit da região.

### 2.1.2 Once its bought when it's the best time period to sell it and for what price?

### 2.1.3 To rise the housing selling price, the company should do a renovation. So what would be good renewal changes?

Now to understand what would be good renewal changes, we must have a discretization of the house total size in m2, to group the houses respectively, and compare the houses that were sold with a price/m2 below the median of its region and now below the median of this total size group median price, what are the amenities that must be added or renewed to increase its market appreciation. 

This way a new categorical variable must represent the discretization of the properties sizes within each region and get a more normal distribution of property size. 

Then calculate the respective median price/m2 for each sub group on each region, and do a visualization to compare the amenities of the houses from each region within its respective property size interval that are below the median price/m2 against the houses that are above, and finally find what must be renewed or built. 

## 3. Set new business hypothesis
***

1. Houses that have a water view, are 20% more expensive on average.


2. Houses with year built older than 1955, are 50% cheaper on average.


3. Houses without basement are 40% bigger than with basement.


4. The growth rate of the houses price YoY (Year over Year) is 10%.


5. Houses with 3 bathrooms have a MoM (Month over Month) growth of 15%.


6. Houses in the mountains are valued 10% more in the Summer than in the Winter.


7. Houses with sqft_living

### 3.1 Test New Business Hypothesis
***

In [ ]:
house_df_eda['renovated'] = house_df_eda['yr_renovated'].apply(lambda x: 1 if x!=0 else 0)
house_df_eda['renovated'].value_counts()

In [ ]:
house_df_eda['region'] = house_df_eda['lat'].apply(lambda x: 'North' if x > 47.46 else 'South')

sampleNorth = house_df_eda[house_df_eda['region'] == 'North'].sample(100)
sampleSouth = house_df_eda[house_df_eda['region'] == 'South'].sample(100)

house_df_sampled = pd.concat([sampleNorth,sampleSouth])

house_df_sampled['price_cube'] = house_df_sampled['price']**1.5

In [ ]:
data_map = house_df_sampled[['id','zipcode', 'lat', 'long', 'price','bedrooms','renovated','region','price_cube']]

map = px.scatter_mapbox(data_map, lat='lat', lon='long',
                       hover_name='id',
                       hover_data=['price'],
                       size= 'price_cube',
                       color='region',
                       color_discrete_sequence=['red','green'],
                       size_max=30,
                       zoom=10,
                       height=300)

map.update_layout(mapbox_style='open-street-map')
map.update_layout(height=600, margin={'r':0, 't':0, 'l':0, 'b':0})

map.show()